module/VuFind/src/VuFind/RateLimiter/Turnstile/Turnstile.php

```diff
@@ -0,0 +1,181 @@
+<?php
+
+/**
+ * Cloudflare Turnstile service.
+ *
+ * PHP version 8
+ *
+ * Copyright (C) Villanova University 2024.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  RateLimiter
+ * @author   Maccabee Levine <msl321@lehigh.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Page
+ */
+
+namespace VuFind\RateLimiter\Turnstile;
+
+use Laminas\Cache\Storage\StorageInterface;
+use Laminas\Log\LoggerAwareInterface;
+use Laminas\Mvc\MvcEvent;
+use VuFind\Log\LoggerAwareTrait;
+use VuFindHttp\HttpServiceAwareInterface;
+use VuFindHttp\HttpServiceAwareTrait;
+
+/**
+ * Rate limiter manager.
+ *
+ * @category VuFind
+ * @package  RateLimiter
+ * @author   Maccabee Levine <msl321@lehigh.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Page
+ */
+class Turnstile implements HttpServiceAwareInterface, LoggerAwareInterface
+{
+    use HttpServiceAwareTrait;
+    use LoggerAwareTrait;
+
+    /**
+     * Constructor
+     *
+     * @param array            $config         Rate Limiter config
+     * @param StorageInterface $turnstileCache Cache for Turnstile results
+     */
+    public function __construct(
+        protected array $config,
+        protected StorageInterface $turnstileCache,
+    ) {
+    }
+
+    /**
+     * Determines if a Turnstile challenge is allowed based on the current event.
+     *
+     * @param MvcEvent $event The MVC event
+     *
+     * @return bool Whether or not the challenge is allowed
+     */
+    public function isChallengeAllowed($event)
+    {
+        $routeMatch = $event->getRouteMatch();
+        $controller = $routeMatch?->getParam('controller') ?? '??';
+        if ('AJAX' === $controller || str_contains($controller, 'Api')) {
```

Comment:
On @EreMaijala 's good point I don't want to present the challenge on AJAX or API requests.  I don't love the str_contains here but there doesn't seem to be any better way of identifying those routes.

Reply:
Good call, done.